be/src/exprs/fpe.cpp

```diff
@@ -0,0 +1,206 @@
+        result_len += dec_part_len;
+
+        result[result_len] = FIXED_NUM;
+        ++result_len;
+    }
+
+    result.resize(result_len);
+    value = result;
+
+    return Status::OK();
+}
+
+Status FPE::decrypt(const std::string_view& num_str, const std::string_view& key, std::string& value, int radix= 10) {
+    int num_str_length = num_str.length();
+    int fpe_key_length = key.length();
+
+    std::vector<uint8_t> fpe_key(fpe_key_length);
+    for (size_t i = 0; i < fpe_key_length; ++i) {
+        fpe_key[i] = static_cast<uint8_t>(key[i]);
+    }
+
+    struct ff1_ctx* ctx = nullptr;
+    DeferOp op([&]{ if (ctx != nullptr) ff1_ctx_destroy(ctx);});
+
+    std::string tmp;
+    tmp.resize(num_str_length);
+    char* out = tmp.data();
+
+    int res = ff1_ctx_create(&ctx, fpe_key.data(),
+                             fpe_key_length, TWEAK, sizeof(TWEAK),
+                             0, SIZE_MAX,
+                             radix);
+    if (res != 0) {
+        return Status::RuntimeError("ff1_ctx_create failed");
+    }
+    res = ff1_decrypt(ctx, out, num_str.data(), NULL, 0);
+    if (res != 0) {
+        return Status::RuntimeError("ff1_encrypt failed");
+    }
+
+    std::string result(out);
+    value = result;
+
+    return Status::OK();
+}
+
+Status FPE::decrypt_num(const std::string_view& num_str, const std::string_view& key, std::string& value) {
+    std::string num_flag;
+    std::string decrypt_num_str(num_str);
+    if (decrypt_num_str[0] == '-') {
+        num_flag = "-";
+        decrypt_num_str = decrypt_num_str.substr(1);
+    }
+
+    Status status;
+    std::string encrypted_dec_part;
+    size_t dot_pos = decrypt_num_str.find('.');
+    std::string int_part = decrypt_num_str.substr(1, dot_pos - 1);
+    status = decrypt(int_part, key, encrypted_dec_part, DEFAULT_RADIX);
+    if (!status.ok()) {
+        return Status::RuntimeError("decrypt_num int_part failed");
+    }
+    std::string_view decrypted_int_part = trim_leading_zeros(encrypted_dec_part);
+
+    std::string decrypted_dec_part;
+    if (dot_pos != std::string::npos) {
+        std::string dec_part = decrypt_num_str.substr(dot_pos + 1, decrypt_num_str.length() - dot_pos - 2);
+        status = decrypt(dec_part, key, decrypted_dec_part, DEFAULT_RADIX);
+        if (!status.ok()) {
+            return Status::RuntimeError("decrypt_num dec_part failed");
+        }
+        int expanded_length = std::to_string(static_cast<long long>(EXPANDED)).length() - 1;
+        std::string leading_zeros(expanded_length - decrypted_dec_part.length(), '0');
+        decrypted_dec_part = leading_zeros + decrypted_dec_part;
+        decrypted_dec_part = trim_trailing_zeros(decrypted_dec_part);
+        decrypted_dec_part = '.' + decrypted_dec_part;
+    }
+
+    std::string decrypted_num = num_flag + std::string(decrypted_int_part) + decrypted_dec_part;
+    value = decrypted_num;
```

Comment:
`value = decrypted_num` need copy the string from `decrypted_num` to `value`.
```suggestion
    value = num_flag + std::string(decrypted_int_part) + decrypted_dec_part;
```